Add a linear interface to jni

facundominguez opened the pull request.

mboes commented on jni/jni.cabal:
```diff
@@ -48,9 +46,16 @@ library
       Data.Singletons
+      Foreign.JNI.Linear
+      Foreign.JNI.Types.Linear
+      Linear.Extras
```
These additions to `linear-base` should be upstreamed. And meanwhile should not be exported to the user, since the module name will sure clash.

facundominguez commented:
All good comments. Thanks for the review.

facundominguez commented on jni/src/linear-types/Foreign/JNI/Linear.hs:
```diff
@@ -0,0 +1,472 @@
+-- of JNI. All local references to objects are tracked so the compiler
+-- ensures that they are eventually deleted and reasonably fast.
+--
+-- Class references and global references aren't tracked because their
```
Class references are not global. We can state that explicitly, indeed.  The call methods could return class references, and it wouldn't be trivial to discriminate them to make them global.

facundominguez replied:
On second thought, maybe we could have our class producing functions `findClass` and `getObjectClass` produce global references. If the call methods return a class, the caller will know it is not global because it has a different type.

facundominguez commented:
@mboes I think I addressed all comments.

I have renamed Linear.hs modules to Safe.hs, but I kept JNI.hs as is. I would rename it to unsafe when there is a GHC release that can build the Safe modules. But I don't feel strong about it.

mboes commented:
Well, as I said previously, you can have both. Move the module to `Unsafe.hs`, then add a `JNI.hs` that simply reexports that module. That way no breaking changes for the user, in particular on a compiler that doesn't have support for linear types.

facundominguez commented:
It was not a matter of how to do it. Anyway, I've create the Unsafe module now.
